kv/range_cache.go

```diff
@@ -108,18 +110,20 @@ func (rmc *rangeDescriptorCache) String() string {
 // the key's data, or an error if any occurred.
 func (rmc *rangeDescriptorCache) LookupRangeDescriptor(key proto.Key,
 	options lookupOptions) (*proto.RangeDescriptor, error) {
-	_, r := rmc.getCachedRangeDescriptor(key)
-	if r != nil {
+	if _, r := rmc.getCachedRangeDescriptor(key); r != nil {
 		return r, nil
 	}
+
 	if log.V(1) {
-		log.Infof("lookup range descriptor: key=%s desc=%+v\n%s", key, r, rmc)
+		log.Infof("lookup range descriptor: key=%s\n%s", key, rmc)
 	}
 
-	rs, err := rmc.db.getRangeDescriptor(key, options)
+	rs, err := rmc.db.getRangeDescriptors(key, options)
 	if err != nil {
 		return nil, err
 	}
+	// TODO(tamird): there is a race here; multiple readers may experience cache
+	// misses and concurrently attempt to refresh the cache, duplicating work.
 	rmc.rangeCacheMu.Lock()
```

Comment:
// Locking over the getRangeDescriptors call is even worse though, given that that blocks the cache completely for the duration of a slow query to the cluster.